### ISSUE TITLE
Materioformer

### ISSUE THREAD

**Issue by JamesStrabala** [JamesStrabala/Capstone-Resorces#2]
### Title
Generative pretrained autoregressive transformer graph neural network applied to the analysis and discovery of novel proteins

###[URL](https://www.science.org/doi/10.1126/science.ads0018](https://pubs.aip.org/aip/jap/article/134/8/084902/2908328/Generative-pretrained-autoregressive-transformer)

### Summary
MaterioFormer integrates a transformer and a graph to predict secondary structure, sequence, and solubility of a protein. It can also generate them to match prompts.  It uses "casual multi-headed graph self-attention" It performs dot-product attention that is then added to the result of a graph. (note:I need to learn how to interpret stuff like the graph formula:<math xmlns="http://www.w3.org/1998/Math/MathML"> <mi xmlns="">E</mi> <mo xmlns="">=</mo> <mrow xmlns=""> <mi mathvariant="normal">softmax</mi></mrow> <mrow xmlns=""> <mo>(</mo> <mrow> <mrow> <mrow> <mo stretchy="false">(</mo> <mrow> <mi>Q</mi> <msup> <mi>K</mi> <mi>T</mi></msup> <mo>+</mo> <mi>M</mi></mrow> <mo stretchy="false">)</mo></mrow> <mo>/</mo> <mrow> <msqrt> <msub> <mi>d</mi> <mi>k</mi></msub></msqrt></mrow></mrow></mrow> <mo>)</mo></mrow></math> Github apparently can't display this properly, but the point stands.)
Formula for an edge:
<math xmlns="http://www.w3.org/1998/Math/MathML"> <msub xmlns=""> <mi>V</mi> <mi>i</mi></msub> <mo xmlns="">=</mo> <mi xmlns="">V</mi> <msubsup xmlns=""> <mi>W</mi> <mi>i</mi> <mi>V</mi></msubsup> <mspace width=".1em" xmlns=""></mspace> <mspace width=".1em" xmlns=""></mspace> <mo stretchy="false" xmlns="">(</mo> <mrow xmlns=""> <mo>∈</mo> <msup> <mrow> <mrow> <mrow> <mi mathvariant="double-struck">R</mi></mrow></mrow></mrow> <mrow> <mi>N</mi> <mo>×</mo> <msub> <mi>d</mi> <mrow> <mi>v</mi> <mo>,</mo> <mi>i</mi></mrow></msub></mrow></msup></mrow> <mo stretchy="false" xmlns="">)</mo> <mo xmlns="">.</mo></math> 
The graphs are put into a graph neural net and the result of that is a matrix with the same dimensions as the dot product, which it's added to. There is also a triangular mask used to ensure the algorithm only builds off previous elements.
### Key Points 
* Single direction transformer that can predict the attributes of a sequence or generate a new one to fit a prompt. 
* Uses a graph to supplement dot-product attention.
### Citation
Buehler, M. J. (2023). Generative pretrained autoregressive transformer graph neural network applied to the analysis and discovery of novel proteins. Journal of Applied Physics, 134(8). https://doi.org/10.1063/5.0157367
@ARTICLE{Buehler2023-gz,
  title     = "Generative pretrained autoregressive transformer graph neural
               network applied to the analysis and discovery of novel proteins",
  author    = "Buehler, Markus J",
  abstract  = "We report a flexible language-model-based deep learning
               strategy, applied here to solve complex forward and inverse
               problems in protein modeling, based on an attention neural
               network that integrates transformer and graph convolutional
               architectures in a causal multi-headed graph mechanism, to
               realize a generative pretrained model. The model is applied to
               predict the secondary structure content (per-residue level and
               overall content), protein solubility, and sequencing tasks.
               Further trained on inverse tasks, the model is rendered capable
               of designing proteins with these properties as target features.
               The model is formulated as a general framework, completely
               prompt-based, and can be adapted for a variety of downstream
               tasks. We find that adding additional tasks yields emergent
               synergies that the model exploits in improving overall
               performance, beyond what would be possible by training a model
               on each dataset alone. Case studies are presented to validate
               the method, yielding protein designs specifically focused on
               structural materials, but also exploring the applicability in
               the design of soluble, antimicrobial biomaterials. While our
               model is trained to ultimately perform eight distinct tasks,
               with available datasets, it can be extended to solve additional
               problems. In a broader sense, this study illustrates a form of
               multiscale modeling that relates a set of ultimate building
               blocks (here, byte-level utf8 characters that define the nature
               of the physical system at hand) to complex output. This
               materiomic scheme captures complex emergent relationships
               between universal building block and resulting properties, via a
               synergizing learning capacity, to express a set of
               potentialities embedded in the knowledge used in training via
               the interplay of universality and diversity. Significance
               statement: Predicting the properties of materials based on a
               flexible description of their structure, environment, or
               process, is a long-standing challenge in multiscale modeling.
               Our MaterioFormer language model, trained to solve forward and
               inverse tasks, incorporates a deep learning capacity through
               attention and graph strategies to yield a multimodal approach to
               model and design materials. Since our model is prompt-based and
               information is encoded consistently via byte-level utf8
               tokenization, it can process diverse modalities of information,
               such as sequence data, description of tasks, and numbers, and
               offers a flexible workflow that integrates human intelligence
               and artificial intelligence. Autoregressive training, using
               pre-training against a large unlabeled dataset, allows for
               straightforward adjustment of specific objectives.",
  journal   = "J. Appl. Phys.",
  publisher = "AIP Publishing",
  volume    =  134,
  number    =  8,
  month     =  aug,
  year      =  2023,
  language  = "en"
}
### Repo
https://github.com/lamm-mit/MateriomicTransformer
